src/agent/proxy.c

```diff
@@ -0,0 +1,135 @@
+                _cleanup_proxy_job_ ProxyJob *proxy = (ProxyJob *) userdata;
+                r = sd_bus_reply_method_errorf(
+                                proxy->request_message, SD_BUS_ERROR_FAILED, "Proxy service failed to start");
+        }
+
+        if (r < 0) {
+                hirte_log_errorf("Failed to sent ready status message to proxy: %s", strerror(-r));
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Failed to set ready status");
+        }
+
+        return sd_bus_reply_method_return(m, "");
+}
+
+
+ProxyJob *
+                proxy_job_new(Agent *agent,
+                              const char *node,
+                              const char *unit,
+                              const char *mode,
+                              sd_bus_message *request_message) {
+        static uint32_t next_id = 0;
+
+        _cleanup_proxy_job_ ProxyJob *proxy = malloc0(sizeof(ProxyJob));
+        if (proxy == NULL) {
+                return NULL;
+        }
+
+        proxy->ref_count = 1;
+        proxy->id = ++next_id;
+        proxy->agent = agent_ref(agent);
+        proxy->request_message = sd_bus_message_ref(request_message);
+        LIST_INIT(proxy_jobs, proxy);
+
+        proxy->node_name = strdup(node);
+        if (proxy->node_name == NULL) {
+                return NULL;
+        }
+        proxy->unit_name = strdup(unit);
+        if (proxy->unit_name == NULL) {
+                return NULL;
+        }
+        proxy->mode = strdup(mode);
+        if (proxy->mode == NULL) {
+                return NULL;
+        }
+
+        int r = asprintf(&proxy->object_path, "%s/%u", INTERNAL_PROXY_OBJECT_PATH_PREFIX, proxy->id);
+        if (r < 0) {
+                return NULL;
+        }
+
+        return steal_pointer(&proxy);
+}
+
+ProxyJob *proxy_job_ref(ProxyJob *proxy) {
+        proxy->ref_count++;
+        return proxy;
+}
+
+void proxy_job_unref(ProxyJob *proxy) {
+        proxy->ref_count--;
+        if (proxy->ref_count != 0) {
+                return;
+        }
+
+        LIST_REMOVE(proxy_jobs, proxy->agent->outstanding_proxy_jobs, proxy);
+
+        sd_bus_message_unrefp(&proxy->request_message);
+        sd_bus_slot_unrefp(&proxy->export_slot);
+
+        free(proxy->node_name);
+        free(proxy->unit_name);
+        free(proxy->mode);
+        free(proxy->object_path);
+        free(proxy);
+}
+
+bool proxy_job_export(ProxyJob *proxy) {
+        int r = sd_bus_add_object_vtable(
+                        proxy->agent->peer_dbus,
+                        &proxy->export_slot,
+                        proxy->object_path,
+                        INTERNAL_PROXY_INTERFACE,
+                        proxy_job_vtable,
+                        proxy);
+        if (r < 0) {
+                hirte_log_errorf("Failed to add service proxy vtable: %s", strerror(-r));
+                return false;
+        }
+        return true;
+}
+
+
+int proxy_job_emit_proxy_new(ProxyJob *proxy) {
```

Comment:
These emit functions should check that `agent->connection_state == AGENT_CONNECTION_STATE_CONNECTED`, or otherwise return some error. Otherwise we might try to send a message on peer_dbus during the connection cycle.

As I said above, if we are disconnected when this is called, we will call it again when connecting.